Add segment module (#3)
WayScience/cell-health-data: 1.segment-data/README.md

# Segment Cell Painting Images

In this module, we present our pipeline for segmenting nuclei and cyoplasm from the Cell Health data.

### Segmentation

> We use the CellPose segmentation algorithim to segment the nuclei from each mitosis movie. 

Reviewer: ```suggestion
We use the CellPose segmentation algorithm to segment the nuclei from the Hoechst channel of each Cell Painting image. 
```

CellPose was first introduced in [Stringer, C., Wang, T., Michaelos, M. et al., 2020](https://doi.org/10.1038/s41592-020-01018-x) and we use the [python implementation](https://github.com/mouseland/cellpose).

Stringer et al. trained the CellPose segmentation models on a diverse set of cell images and is therefore a good selection for our use case.
The CellPose python implementation was particularly useful for building reproducible pipelines.

> After manually experiementing with CellPose on about 10 Cell Health nuclei images (various cell lines), we settled on the following parameters for CellPose nuclei segmentation:

Reviewer: ```suggestion

### Nucleus segmentation

After manually experiementing with CellPose on about 10 Cell Health nuclei images (various cell lines), we settled on the following parameters for CellPose nuclei segmentation:
```

Just to clearly delineate sections

- `model_type = "cyto"` This parameter forces CellPose to use the cytoplasm model, which we found segments nuclei in the Cell Health data significantly better than the nucleus model. 
More information about CellPose models can be found at https://cellpose.readthedocs.io/en/latest/models.html.
- `channels = [0,0]` This parameter forces the model to segment cells in grayscale (in the case of Cell Health data single channel images).

> - `diameter = 80` This parameter indicates to the model that the average cell diameter is 80 pixels.

Reviewer: Is this the same for all cell lines? IIRC, ES2 cells are larger. (but mayeb this parameter doesn't matter too much)

Author: Yes, these are the same for all cell lines. We manually testing these configurations for all 3 cell lines. This parameter doesn't matter too much, mostly just ensures the cells don't get over-segmented (produce very small nuclei segmentation).

- `flow_threshold=0` This paramenter decreases the maximum allowed error of the flows for each mask (default is `flow_threshold=0.4`).
More information about CellPose settings can be found at https://cellpose.readthedocs.io/en/latest/settings.html.

> After manually experiementing with CellPose on about 10 Cell Health cytoplasm images (various cell lines), we settled on the following parameters for CellPose cytoplasm segmentation:

Reviewer: ```suggestion
### Cytoplasm segmentation

After manually experimenting with CellPose on about 10 Cell Health cytoplasm images (various cell lines), we settled on the following parameters for CellPose cytoplasm segmentation:
```

Fixing a typo and to clearly delineate section

- `model_type = "cyto"` This parameter forces CellPose to use the cytoplasm model for segmentation.
More information about CellPose models can be found at https://cellpose.readthedocs.io/en/latest/models.html.

> - `channels = [1,3]` This parameter forces the model to segment cells using the nuclei and RNA channels of cell painting images. We overlay nuclei, ER, and RNA channels for CellPose with our `overlay_images()` function in [segment-cell-health-data.ipynb](segment-cell-health-data.ipynb).

Reviewer: do you mean channels = [1, 2, 3]? I see three channels here, but maybe I don't understand this parameter!

Reviewer 2: In CellPose, the first part of `channels` represents the channel being segmented which is the RNA/cytoplasm channel. The second part is the nucleus channel that helps CellPose with segmentation. The screenshot below shows what it looks like but it is currently set for the NF1 data. Note that the channels in CellPose is based on the color so even though for the cell health data, channel 1 is nuclei, it is channel 3 because it is blue. 
![cellpose](https://user-images.githubusercontent.com/107513215/179846992-52bd27f0-ce27-4f90-84af-8963d6c767da.png)

- `flow_threshold=0` This paramenter decreases the maximum allowed error of the flows for each mask (default is `flow_threshold=0.4`).
- `cellprob_threshold=0.4` This parameter is used to determine ROIs (default is `cellprob_threshold=0`). More information about CellPose settings can be found at https://cellpose.readthedocs.io/en/latest/settings.html.

## Step 1: Setup Segmentation Environment

### Step 1a: Create Segmentation Environment

```sh
# Run this command to create the conda environment for Segmentation data
conda env create -f 1.segment-env.yml
```

### Step 1b: Activate Segmentation Environment

```sh
# Run this command to activate the conda environment for Segmentation data
conda activate 1.segment-data-cell-health
```

### Step 1c (Optional): Complete PyTorch GPU Setup

If you would like use PyTorch GPU when using CellPose, follow [these instructions](https://github.com/MouseLand/cellpose#gpu-version-cuda-on-windows-or-linux) to complete the PyTorch GPU setup.
We use PyTorch GPU while segmenting Cell Health data.

## Step 2: Execute Training Data Segmentation

```bash
# Run this script to segment Cell Health data
bash 1.segment-data.sh

> ```

Reviewer: Can you add a comment somewhere here regarding:

1. Your CPU/GPU specs
2. The estimated run time for each segmentation model

Author: Commenting these times so I don't forget:
nuclei - 18 hrs
cytoplasm - 58 hrs
